pydis_site/apps/api/tests/test_reminders.py

```diff
@@ -0,0 +1,139 @@
+from datetime import datetime
+
+from django_hosts.resolvers import reverse
+
+from .base import APISubdomainTestCase
+from ..models import Reminder, User
+
+
+class UnauthedReminderAPITests(APISubdomainTestCase):
+    def setUp(self):
+        super().setUp()
+        self.client.force_authenticate(user=None)
+
+    def test_list_returns_401(self):
+        url = reverse('bot:reminder-list', host='api')
+        response = self.client.get(url)
+
+        self.assertEqual(response.status_code, 401)
+
+    def test_create_returns_401(self):
+        url = reverse('bot:reminder-list', host='api')
+        response = self.client.post(url, data={'not': 'important'})
+
+        self.assertEqual(response.status_code, 401)
+
+    def test_delete_returns_401(self):
+        url = reverse('bot:reminder-detail', args=('1234',), host='api')
+        response = self.client.delete(url)
+
+        self.assertEqual(response.status_code, 401)
+
+
+class EmptyDatabaseReminderAPITests(APISubdomainTestCase):
+    def test_list_all_returns_empty_list(self):
+        url = reverse('bot:reminder-list', host='api')
+        response = self.client.get(url)
+
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(response.json(), [])
+
+    def test_delete_returns_404(self):
+        url = reverse('bot:reminder-detail', args=('1234',), host='api')
+        response = self.client.delete(url)
+
+        self.assertEqual(response.status_code, 404)
+
+
+class ReminderCreationTests(APISubdomainTestCase):
```

Comment:
This lacks tests for invalid data.

Reply:
True. Would you suggest a separate test for each field that can be invalid? I'm particularly weak on testing and trying to learn best practices.

Reply:
I think a single test for _anything_ invalid is adequate to ensure the API reacts appropriately to erroneous data.

Here's my perspective on why:

Testing each field may be more applicable to model tests (which we don't do anywhere I believe, but maybe we should...), but these are API/viewset tests. The viewset is using standard mixins/base classes from DRF without any overrides. Those mixins are tested by DRF themselves. If the viewset used custom mixins and/or custom method handlers, then it would be a different situation regarding testing.

Reply:
That's very helpful, I lost sight of that context regarding the DRF mixins and their existing tests.